CRM/ManualDirectDebit/Form/Batch.php

```diff
@@ -124,4 +123,29 @@ private function createBatchActivity($batchTitle) {
     CRM_Activity_BAO_Activity::create($activityParams);
   }
 
+  /**
+   * Intercept QF validation and do our own redirection.
+   *
+   * @return bool
+   *   true if no error found
+   */
+  public function validate() {
+    $errors = parent::validate();
+    $params = $this->controller->exportValues($this->_name);
+
+    if ($params['batch_id'] <= CRM_ManualDirectDebit_Batch_BatchHandler::getMaxBatchId()) {
+      $title = ts('Batch ID already exists');
+      $message = ts('Batch ID %1 already exists. It is likely another batch 
+        has just been created a moment ago.', [1 => $params['batch_id']]);
+
+      CRM_Core_Session::setStatus($message, $title, 'error');
+
+      $this->controller->setDestination(NULL, TRUE);
+      header("Refresh:0; {$_SERVER['HTTP_REFERER']}");
+      exit;
```

Comment:
We generally try to avoid things like this if possible (that is, forcibly redirecting and exiting, breaking execution), unless it's absolutely necessary. What we talked about yesterday was just doing a normal validation (with a rule for the quickform), so that if the validation failed, then we'd just show the message suggested by @guanhuan, "Batch ID %1 already exists. It is likely another batch has just been created a moment ago. Please refresh the page and try again."

If there is a reason why you chose to redirect and exit execution, can you please explain? Otherwise, we would really prefer to just fail the rule, show the normal validation message and ask the user himself to refresh the page.

Eg.
```php
public function buildQuickForm() {
  parent::buildQuickForm();
  $this->addFormRule(array(self::class, 'validateBatchID'));
}

public function valiateBatchID() {
  $params = $this->controller->exportValues($this->_name);

  if ($params['batch_id'] <= CRM_ManualDirectDebit_Batch_BatchHandler::getMaxBatchId()) {
    $errors['batch_id'] = ts('Batch ID %1 already exists. It is likely another batch has just been created a moment ago.', [1 => $params['batch_id']]);
  }

  return empty($errors) ? TRUE : $errors;
}
```

Reply:
The GET Parameters disappears when the errors are detected, so the form is not working after refreshing the page. It is necessary to go back to the link in the menu.
That is the main reason we choose to redirect and exit execution.

Reply:
@awg-github 

I'm not sure I understand the problem. What GET parameters disappear?

Reply:
@awg-github

I'm not sure I understand the problem. What GET parameters disappear?